server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ChunkedStreamWithSize.java

```diff
@@ -0,0 +1,150 @@
+ */
+public class ChunkedStreamWithSize implements ChunkedInput<ByteBuf> {
+
+    static final int DEFAULT_CHUNK_SIZE = 8192;
+
+    private final PushbackInputStream in;
+    private final int chunkSize;
+    private long offset;
+    private boolean closed;
+
+    /**
+     * Creates a new instance that fetches data from the specified stream.
+     */
+    public ChunkedStreamWithSize(InputStream in) {
+        this(in, DEFAULT_CHUNK_SIZE);
+    }
+
+    /**
+     * Creates a new instance that fetches data from the specified stream.
+     *
+     * @param chunkSize the number of bytes to fetch on each
+     *                  {@link #readChunk(ChannelHandlerContext)} call
+     */
+    public ChunkedStreamWithSize(InputStream in, int chunkSize) {
+        ObjectUtil.checkNotNull(in, "in");
+        ObjectUtil.checkPositive(chunkSize, "chunkSize");
+
+        if (in instanceof PushbackInputStream) {
+            this.in = (PushbackInputStream) in;
+        } else {
+            this.in = new PushbackInputStream(in);
+        }
+        this.chunkSize = chunkSize;
+    }
+
+    /**
+     * Returns the number of transferred bytes.
+     */
+    public long transferredBytes() {
+        return offset;
+    }
+
+    @Override
+    public boolean isEndOfInput() throws Exception {
+        if (closed) {
+            return true;
+        }
+        if (in.available() > 0) {
+            return false;
+        }
+
+        int b = in.read();
+        if (b < 0) {
+            return true;
+        } else {
+            in.unread(b);
+            return false;
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        closed = true;
+        in.close();
+    }
+
+    @Deprecated
+    @Override
+    public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+        return readChunk(ctx.alloc());
+    }
+
+    @Override
+    public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+        if (isEndOfInput()) {
+            return null;
+        }
+
+        final int availableBytes = in.available();
+        final int chunkSize;
+        if (availableBytes <= 0) {
+            chunkSize = this.chunkSize;
+        } else {
+            chunkSize = Math.min(this.chunkSize, in.available());
+        }
+
+        boolean release = true;
+        ByteBuf buffer = allocator.buffer(chunkSize);
+        try {
+            // transfer to buffer
+            int written = buffer.writeBytes(in, chunkSize);
+            if (written < 0) {
+                return null;
+            }
+            offset += written;
+            release = false;
+            return buffer;
+        } finally {
+            if (release) {
+                buffer.release();
+            }
+        }
+    }
+
+    @Override
+    public long length() {
+        return -1;
+    }
+
+    @Override
+    public long progress() {
+        return offset;
+    }
+
+    public int getChunkSize() {
+        return chunkSize;
+    }
```

Comment:
Well James do actually chucks some kinda transformations of "ByteArrayInputStream".

So the chuck size is NOT representative of overall consummed memory by the write task.